# Pandas Tutorial

In [1]:
# Import pandas
import pandas as pd

Creating Pandas Dataframes
--

In [15]:
# Creating Pandas Dataframe
create_df = pd.DataFrame({'A':[1,'11'], 'B':[2,'22'], 'C':[3,'33'], 'D':[4,'44']})

In [16]:
# View the created Dataframe
create_df

A   B   C   D
0   1   2   3   4
1  11  22  33  44

In [19]:
# Creating Pandas Dataframe with row index names
sample_create_df = pd.DataFrame({'A':[1,'11'], 'B':[2,'22'], 'C':[3,'33'], 'D':[4,'44']}, index=['Row0', 'Row1'])

In [20]:
# View created dataframe
sample_create_df

A   B   C   D
Row0   1   2   3   4
Row1  11  22  33  44

Creating Pandas Series
--

In [21]:
# Creating a pandas Series
sample_series = pd.Series(['A',1,'B',2,'C',3])

In [22]:
# View created series
sample_series

0    A
1    1
2    B
3    2
4    C
5    3
dtype: object

In [24]:
# Creating a pandas series with a name and row index name
sample_series_with_name = pd.Series(['A',1,'B',2,'C',3], index=['Row0','Row1','Row2','Row3','Row4', 'Row5'], name='Sample Pandas Series')

In [25]:
# View created series
sample_series_with_name

Row0    A
Row1    1
Row2    B
Row3    2
Row4    C
Row5    3
Name: Sample Pandas Series, dtype: object

Reading from a file using Pandas
--

In [3]:
# Read from a CSV file
readd = pd.read_csv('E:/Data Wrangling/Prac_Assignment/Assignment2/Project_Assignment2/au.csv')

In [5]:
# View first few rows
readd.head()

city      lat       lng    country iso2         admin_name capital  \
0     Sydney -33.8650  151.2094  Australia   AU    New South Wales   admin   
1  Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
2   Brisbane -27.4678  153.0281  Australia   AU         Queensland   admin   
3      Perth -31.9522  115.8589  Australia   AU  Western Australia   admin   
4   Adelaide -34.9289  138.6011  Australia   AU    South Australia   admin   

   population  population_proper  
0     5312163            4840600  
1     5078193            4529500  
2     2514184            2360241  
3     2059484            2039200  
4     1345777            1295714

In [6]:
# Find shape (rows x columns)
readd.shape

(1035, 9)

In [7]:
# View last few rows
readd.tail()

city      lat       lng    country iso2       admin_name capital  \
1030  The Palms -26.1902  152.5861  Australia   AU       Queensland     NaN   
1031  Millfield -32.8886  151.2614  Australia   AU  New South Wales     NaN   
1032   Dunbogan -31.6500  152.8167  Australia   AU  New South Wales     NaN   
1033    Tomakin -35.8228  150.1894  Australia   AU  New South Wales     NaN   
1034     Currie -39.9311  143.8510  Australia   AU         Tasmania   minor   

      population  population_proper  
1030        1010               1010  
1031        1006               1006  
1032        1002               1002  
1033        1001               1001  
1034         768                768

In [8]:
# Read from csv file (from col 0)
readd0 = pd.read_csv('E:/Data Wrangling/Prac_Assignment/Assignment2/Project_Assignment2/au.csv', index_col=0)

In [9]:
# View first few rows
readd0.head()

lat       lng    country iso2         admin_name capital  \
city                                                                      
Sydney    -33.8650  151.2094  Australia   AU    New South Wales   admin   
Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
Brisbane  -27.4678  153.0281  Australia   AU         Queensland   admin   
Perth     -31.9522  115.8589  Australia   AU  Western Australia   admin   
Adelaide  -34.9289  138.6011  Australia   AU    South Australia   admin   

           population  population_proper  
city                                      
Sydney        5312163            4840600  
Melbourne     5078193            4529500  
Brisbane      2514184            2360241  
Perth         2059484            2039200  
Adelaide      1345777            1295714

In [10]:
# Find shape (rows x columns)
readd0.shape

(1035, 8)

In [11]:
# View last few rows
readd0.tail()

lat       lng    country iso2       admin_name capital  \
city                                                                    
The Palms -26.1902  152.5861  Australia   AU       Queensland     NaN   
Millfield -32.8886  151.2614  Australia   AU  New South Wales     NaN   
Dunbogan  -31.6500  152.8167  Australia   AU  New South Wales     NaN   
Tomakin   -35.8228  150.1894  Australia   AU  New South Wales     NaN   
Currie    -39.9311  143.8510  Australia   AU         Tasmania   minor   

           population  population_proper  
city                                      
The Palms        1010               1010  
Millfield        1006               1006  
Dunbogan         1002               1002  
Tomakin          1001               1001  
Currie            768                768

Accessing property from a Dataframe
---

In [26]:
readd.city

0          Sydney
1       Melbourne
2        Brisbane
3           Perth
4        Adelaide
          ...    
1030    The Palms
1031    Millfield
1032     Dunbogan
1033      Tomakin
1034       Currie
Name: city, Length: 1035, dtype: object

In [27]:
readd['city']

0          Sydney
1       Melbourne
2        Brisbane
3           Perth
4        Adelaide
          ...    
1030    The Palms
1031    Millfield
1032     Dunbogan
1033      Tomakin
1034       Currie
Name: city, Length: 1035, dtype: object

In [28]:
# Access first element of the property
readd['city'][0]

'Sydney'

Indexing in Pandas - iloc (passing numerical position as index)
---

In [29]:
# Accessing all columns and 0th row only
readd.iloc[0]

city                          Sydney
lat                          -33.865
lng                         151.2094
country                    Australia
iso2                              AU
admin_name           New South Wales
capital                        admin
population                   5312163
population_proper            4840600
Name: 0, dtype: object

In [30]:
# Accessing all rows and 0th column only
readd.iloc[:,0]

0          Sydney
1       Melbourne
2        Brisbane
3           Perth
4        Adelaide
          ...    
1030    The Palms
1031    Millfield
1032     Dunbogan
1033      Tomakin
1034       Currie
Name: city, Length: 1035, dtype: object

In [31]:
# Accessing 1st and 2nd row of 0th column
readd.iloc[1:3,0]

1    Melbourne
2     Brisbane
Name: city, dtype: object

In [32]:
# Accessing 1st and 2nd row of 0,1,2 columns
readd.iloc[1:3,0:3]

city      lat       lng
1  Melbourne -37.8136  144.9631
2   Brisbane -27.4678  153.0281

Indexing in Pandas - loc (passing label as index)
------------------------

In [33]:
# Accessing 0th element from readd dataframe
readd.loc[0,'city']

'Sydney'

In [35]:
# Accessing all rows from list of columns of readd dataframe
readd.loc[:, ['city','country','admin_name']]

city    country         admin_name
0        Sydney  Australia    New South Wales
1     Melbourne  Australia           Victoria
2      Brisbane  Australia         Queensland
3         Perth  Australia  Western Australia
4      Adelaide  Australia    South Australia
...         ...        ...                ...
1030  The Palms  Australia         Queensland
1031  Millfield  Australia    New South Wales
1032   Dunbogan  Australia    New South Wales
1033    Tomakin  Australia    New South Wales
1034     Currie  Australia           Tasmania

[1035 rows x 3 columns]

Choosing between loc and iloc:
---------------------------------------------

When choosing or transitioning between loc and iloc, there is one "gotcha" worth keeping in mind, which is that the two methods use slightly different indexing schemes.

iloc uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So 0:10 will select entries 0,...,9. loc, meanwhile, indexes inclusively. So 0:10 will select entries 0,...,10.

Why the change? Remember that loc can index any stdlib type: strings, for example. If we have a DataFrame with index values Apples, ..., Potatoes, ..., and we want to select "all the alphabetical fruit choices between Apples and Potatoes", then it's a lot more convenient to index df.loc['Apples':'Potatoes'] than it is to index something like df.loc['Apples', 'Potatoet'] (t coming after s in the alphabet).

This is particularly confusing when the DataFrame index is a simple numerical list, e.g. 0,...,1000. In this case df.iloc[0:1000] will return 1000 entries, while df.loc[0:1000] return 1001 of them! To get 1000 elements using loc, you will need to go one lower and ask for df.loc[0:999].


Manipulating the index:
---------------------------------

In [36]:
# Setting the index
readd.set_index('iso2')

city      lat       lng    country         admin_name capital  \
iso2                                                                       
AU       Sydney -33.8650  151.2094  Australia    New South Wales   admin   
AU    Melbourne -37.8136  144.9631  Australia           Victoria   admin   
AU     Brisbane -27.4678  153.0281  Australia         Queensland   admin   
AU        Perth -31.9522  115.8589  Australia  Western Australia   admin   
AU     Adelaide -34.9289  138.6011  Australia    South Australia   admin   
...         ...      ...       ...        ...                ...     ...   
AU    The Palms -26.1902  152.5861  Australia         Queensland     NaN   
AU    Millfield -32.8886  151.2614  Australia    New South Wales     NaN   
AU     Dunbogan -31.6500  152.8167  Australia    New South Wales     NaN   
AU      Tomakin -35.8228  150.1894  Australia    New South Wales     NaN   
AU       Currie -39.9311  143.8510  Australia           Tasmania   minor   

      population  population_proper  
iso2                                 
AU       5312163            4840600  
AU       5078193            4529500  
AU       2514184            2360241  
AU       2059484            2039200  
AU       1345777            1295714  
...          ...                ...  
AU          1010               1010  
AU          1006               1006  
AU          1002               1002  
AU          1001               1001  
AU           768                768  

[1035 rows x 8 columns]

Conditional Selection
---

In [45]:
# To select city names of NSW which have population less than or equal to 2000
cond = readd.loc[(readd.admin_name == 'New South Wales') & (readd.population <= 2000)]

In [46]:
## To view the dataframe
cond

city      lat       lng    country iso2       admin_name  \
628          Nimbin -28.6000  153.2333  Australia   AU  New South Wales   
629       Wahroonga -33.6928  151.1261  Australia   AU  New South Wales   
630   Clarence Town -32.5842  151.7819  Australia   AU  New South Wales   
632        Branxton -32.6483  151.3511  Australia   AU  New South Wales   
633    Camden Haven -31.6483  152.8011  Australia   AU  New South Wales   
...             ...      ...       ...        ...  ...              ...   
1020     Jerilderie -35.3500  145.7333  Australia   AU  New South Wales   
1021    Yellow Rock -33.6983  150.6344  Australia   AU  New South Wales   
1031      Millfield -32.8886  151.2614  Australia   AU  New South Wales   
1032       Dunbogan -31.6500  152.8167  Australia   AU  New South Wales   
1033        Tomakin -35.8228  150.1894  Australia   AU  New South Wales   

     capital  population  population_proper  
628      NaN        2000               2000  
629      NaN        1998               1998  
630      NaN        1994               1994  
632      NaN        1991               1991  
633      NaN        1986               1986  
...      ...         ...                ...  
1020     NaN        1029               1029  
1021     NaN        1029               1029  
1031     NaN        1006               1006  
1032     NaN        1002               1002  
1033     NaN        1001               1001  

[120 rows x 9 columns]

In [47]:
# To get the shape of the dataframe
cond.shape

(120, 9)

In [50]:
# To get the city names of NSW and Vic or city names with population <= 1500
condi = readd.loc[(readd.admin_name.isin(['New South Wales', 'Victoria'])) | readd.population <= 1500]

In [51]:
# To view the dataframe
condi

city      lat       lng    country iso2         admin_name capital  \
0        Sydney -33.8650  151.2094  Australia   AU    New South Wales   admin   
1     Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
2      Brisbane -27.4678  153.0281  Australia   AU         Queensland   admin   
3         Perth -31.9522  115.8589  Australia   AU  Western Australia   admin   
4      Adelaide -34.9289  138.6011  Australia   AU    South Australia   admin   
...         ...      ...       ...        ...  ...                ...     ...   
1030  The Palms -26.1902  152.5861  Australia   AU         Queensland     NaN   
1031  Millfield -32.8886  151.2614  Australia   AU    New South Wales     NaN   
1032   Dunbogan -31.6500  152.8167  Australia   AU    New South Wales     NaN   
1033    Tomakin -35.8228  150.1894  Australia   AU    New South Wales     NaN   
1034     Currie -39.9311  143.8510  Australia   AU           Tasmania   minor   

      population  population_proper  
0        5312163            4840600  
1        5078193            4529500  
2        2514184            2360241  
3        2059484            2039200  
4        1345777            1295714  
...          ...                ...  
1030        1010               1010  
1031        1006               1006  
1032        1002               1002  
1033        1001               1001  
1034         768                768  

[1035 rows x 9 columns]

In [52]:
# To get the shape of the dataframe
condi.shape

(1035, 9)

In [38]:
# Suppose we want to select only the NSW and Victorian cities
cond_1 = readd.loc[(readd.admin_name.isin(['New South Wales','Victoria']))]

In [39]:
# View the dataframe
cond_1

city      lat       lng    country iso2       admin_name  \
0           Sydney -33.8650  151.2094  Australia   AU  New South Wales   
1        Melbourne -37.8136  144.9631  Australia   AU         Victoria   
7        Newcastle -32.9167  151.7500  Australia   AU  New South Wales   
8       Wollongong -34.4331  150.8831  Australia   AU  New South Wales   
9          Geelong -38.1500  144.3500  Australia   AU         Victoria   
...            ...      ...       ...        ...  ...              ...   
1027    Lethbridge -37.9667  144.1330  Australia   AU         Victoria   
1028  Nar Nar Goon -38.0819  145.5750  Australia   AU         Victoria   
1031     Millfield -32.8886  151.2614  Australia   AU  New South Wales   
1032      Dunbogan -31.6500  152.8167  Australia   AU  New South Wales   
1033       Tomakin -35.8228  150.1894  Australia   AU  New South Wales   

     capital  population  population_proper  
0      admin     5312163            4840600  
1      admin     5078193            4529500  
7        NaN      322278             152984  
8        NaN      302739             289236  
9        NaN      263280             191440  
...      ...         ...                ...  
1027     NaN        1014               1014  
1028     NaN        1012               1012  
1031     NaN        1006               1006  
1032     NaN        1002               1002  
1033     NaN        1001               1001  

[567 rows x 9 columns]

In [40]:
# Find the shape of the dataframe
cond_1.shape

(567, 9)

In [42]:
# Suppose we want to select NaN values from capital property
cond_2 = cond_1.loc[(cond_1.capital.isnull())]

# Suppose to get the values which are not null
# cond_3 = cond_1.loc[(cond_1.capital.notnull())]

In [43]:
# View the dataframe
cond_2

city      lat       lng    country iso2       admin_name  \
7        Newcastle -32.9167  151.7500  Australia   AU  New South Wales   
8       Wollongong -34.4331  150.8831  Australia   AU  New South Wales   
9          Geelong -38.1500  144.3500  Australia   AU         Victoria   
16        Ballarat -37.5500  143.8500  Australia   AU         Victoria   
17         Bendigo -36.7500  144.2667  Australia   AU         Victoria   
...            ...      ...       ...        ...  ...              ...   
1027    Lethbridge -37.9667  144.1330  Australia   AU         Victoria   
1028  Nar Nar Goon -38.0819  145.5750  Australia   AU         Victoria   
1031     Millfield -32.8886  151.2614  Australia   AU  New South Wales   
1032      Dunbogan -31.6500  152.8167  Australia   AU  New South Wales   
1033       Tomakin -35.8228  150.1894  Australia   AU  New South Wales   

     capital  population  population_proper  
7        NaN      322278             152984  
8        NaN      302739             289236  
9        NaN      263280             191440  
16       NaN      105471              98543  
17       NaN       99122              95587  
...      ...         ...                ...  
1027     NaN        1014               1014  
1028     NaN        1012               1012  
1031     NaN        1006               1006  
1032     NaN        1002               1002  
1033     NaN        1001               1001  

[565 rows x 9 columns]

In [44]:
# Find the shape of the dataframe
cond_2.shape

(565, 9)

Assigning Data
--

In [53]:
# To create a new column with constant value
readd['newCol'] = 'newVal'

In [54]:
# To view the dataframe
readd

city      lat       lng    country iso2         admin_name capital  \
0        Sydney -33.8650  151.2094  Australia   AU    New South Wales   admin   
1     Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
2      Brisbane -27.4678  153.0281  Australia   AU         Queensland   admin   
3         Perth -31.9522  115.8589  Australia   AU  Western Australia   admin   
4      Adelaide -34.9289  138.6011  Australia   AU    South Australia   admin   
...         ...      ...       ...        ...  ...                ...     ...   
1030  The Palms -26.1902  152.5861  Australia   AU         Queensland     NaN   
1031  Millfield -32.8886  151.2614  Australia   AU    New South Wales     NaN   
1032   Dunbogan -31.6500  152.8167  Australia   AU    New South Wales     NaN   
1033    Tomakin -35.8228  150.1894  Australia   AU    New South Wales     NaN   
1034     Currie -39.9311  143.8510  Australia   AU           Tasmania   minor   

      population  population_proper  newCol  
0        5312163            4840600  newVal  
1        5078193            4529500  newVal  
2        2514184            2360241  newVal  
3        2059484            2039200  newVal  
4        1345777            1295714  newVal  
...          ...                ...     ...  
1030        1010               1010  newVal  
1031        1006               1006  newVal  
1032        1002               1002  newVal  
1033        1001               1001  newVal  
1034         768                768  newVal  

[1035 rows x 10 columns]

In [55]:
# Find the shape of the dataframe
readd.shape

(1035, 10)

In [56]:
# To create a new column with iterating values
readd['newCol1'] = range(len(readd), 0, -1)

In [57]:
# To view the dataframe
readd

city      lat       lng    country iso2         admin_name capital  \
0        Sydney -33.8650  151.2094  Australia   AU    New South Wales   admin   
1     Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
2      Brisbane -27.4678  153.0281  Australia   AU         Queensland   admin   
3         Perth -31.9522  115.8589  Australia   AU  Western Australia   admin   
4      Adelaide -34.9289  138.6011  Australia   AU    South Australia   admin   
...         ...      ...       ...        ...  ...                ...     ...   
1030  The Palms -26.1902  152.5861  Australia   AU         Queensland     NaN   
1031  Millfield -32.8886  151.2614  Australia   AU    New South Wales     NaN   
1032   Dunbogan -31.6500  152.8167  Australia   AU    New South Wales     NaN   
1033    Tomakin -35.8228  150.1894  Australia   AU    New South Wales     NaN   
1034     Currie -39.9311  143.8510  Australia   AU           Tasmania   minor   

      population  population_proper  newCol  newCol1  
0        5312163            4840600  newVal     1035  
1        5078193            4529500  newVal     1034  
2        2514184            2360241  newVal     1033  
3        2059484            2039200  newVal     1032  
4        1345777            1295714  newVal     1031  
...          ...                ...     ...      ...  
1030        1010               1010  newVal        5  
1031        1006               1006  newVal        4  
1032        1002               1002  newVal        3  
1033        1001               1001  newVal        2  
1034         768                768  newVal        1  

[1035 rows x 11 columns]

In [58]:
# To get the shape of the dataframe
readd.shape

(1035, 11)

Summary Functions
--

In [59]:
# describe() - To get more details about a particular property/column. More good for numerical data than string datatype
readd.population.describe()

count    1.035000e+03
mean     2.424799e+04
std      2.545478e+05
min      7.680000e+02
25%      1.521000e+03
50%      2.605000e+03
75%      5.390000e+03
max      5.312163e+06
Name: population, dtype: float64

In [60]:
# unique() - To get unique values from a column
readd.admin_name.unique()

array(['New South Wales', 'Victoria', 'Queensland', 'Western Australia',
       'South Australia', 'Australian Capital Territory', 'Tasmania',
       'Northern Territory'], dtype=object)

In [61]:
# value_counts() - To get the number of times a value is used 
readd.admin_name.value_counts()

New South Wales                 335
Queensland                      240
Victoria                        232
South Australia                  88
Western Australia                83
Tasmania                         46
Northern Territory               10
Australian Capital Territory      1
Name: admin_name, dtype: int64

Maps - To transform from one form to another
--

In [62]:
# Suppose we want to remean the population value
remean_pop = readd.population.mean()

In [63]:
# Using map() to remean the population
readd.population.map(lambda a : a - remean_pop)

0       5.287915e+06
1       5.053945e+06
2       2.489936e+06
3       2.035236e+06
4       1.321529e+06
            ...     
1030   -2.323799e+04
1031   -2.324199e+04
1032   -2.324599e+04
1033   -2.324699e+04
1034   -2.347999e+04
Name: population, Length: 1035, dtype: float64

The function you pass to map() should expect a single value from the Series (a population value, in the above example), and return a transformed version of that value. 
--
map() returns a new Series where all the values have been transformed by your function
--

apply()
--

In [64]:
# Each row values are changed by using a apply()
def remean_pop_fn(row):
    row.population = row.population - remean_pop
    return row

readd.apply(remean_pop_fn, axis='columns')

city      lat       lng    country iso2         admin_name capital  \
0        Sydney -33.8650  151.2094  Australia   AU    New South Wales   admin   
1     Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
2      Brisbane -27.4678  153.0281  Australia   AU         Queensland   admin   
3         Perth -31.9522  115.8589  Australia   AU  Western Australia   admin   
4      Adelaide -34.9289  138.6011  Australia   AU    South Australia   admin   
...         ...      ...       ...        ...  ...                ...     ...   
1030  The Palms -26.1902  152.5861  Australia   AU         Queensland     NaN   
1031  Millfield -32.8886  151.2614  Australia   AU    New South Wales     NaN   
1032   Dunbogan -31.6500  152.8167  Australia   AU    New South Wales     NaN   
1033    Tomakin -35.8228  150.1894  Australia   AU    New South Wales     NaN   
1034     Currie -39.9311  143.8510  Australia   AU           Tasmania   minor   

        population  population_proper  newCol  newCol1  
0     5.287915e+06            4840600  newVal     1035  
1     5.053945e+06            4529500  newVal     1034  
2     2.489936e+06            2360241  newVal     1033  
3     2.035236e+06            2039200  newVal     1032  
4     1.321529e+06            1295714  newVal     1031  
...            ...                ...     ...      ...  
1030 -2.323799e+04               1010  newVal        5  
1031 -2.324199e+04               1006  newVal        4  
1032 -2.324599e+04               1002  newVal        3  
1033 -2.324699e+04               1001  newVal        2  
1034 -2.347999e+04                768  newVal        1  

[1035 rows x 11 columns]

Grouping and Sorting
--

In [65]:
# Replica for value_counts()
readd.groupby('admin_name').admin_name.count()

admin_name
Australian Capital Territory      1
New South Wales                 335
Northern Territory               10
Queensland                      240
South Australia                  88
Tasmania                         46
Victoria                        232
Western Australia                83
Name: admin_name, dtype: int64

In [77]:
# To get min count of population_proper based on admin_name
readd.groupby('admin_name').population_proper.min()

admin_name
Australian Capital Territory    381488
New South Wales                    180
Northern Territory                1081
Queensland                         118
South Australia                    195
Tasmania                           768
Victoria                           865
Western Australia                 1029
Name: population_proper, dtype: int64

In [104]:
# Check the min value of population_proper - 118 is for Queensland?
# Step 1: Get the index of lowest value in population_proper column
readd[['population_proper']].idxmin()

population_proper    74
dtype: int64

In [105]:
# Step 2: View the particular row from the above step
readd.iloc[74:75]

city      lat    lng    country iso2  admin_name capital  population  \
74  Mount Isa -20.7333  139.5  Australia   AU  Queensland     NaN       18588   

    population_proper  newCol  newCol1  
74                118  newVal      961

In [108]:
# To get max count of population_proper based on admin_name
readd.groupby('admin_name').population_proper.max()

admin_name
Australian Capital Territory     381488
New South Wales                 4840600
Northern Territory               136828
Queensland                      2360241
South Australia                 1295714
Tasmania                         222356
Victoria                        4529500
Western Australia               2039200
Name: population_proper, dtype: int64

In [106]:
# Check the max value of population_proper - 4840600 is for NSW?
# Step 1: Get the index of max value in population_proper column
readd[['population_proper']].idxmax()

population_proper    0
dtype: int64

In [107]:
# Step 2: View the particular row from the above step
readd.iloc[0:3]

city      lat       lng    country iso2       admin_name capital  \
0     Sydney -33.8650  151.2094  Australia   AU  New South Wales   admin   
1  Melbourne -37.8136  144.9631  Australia   AU         Victoria   admin   
2   Brisbane -27.4678  153.0281  Australia   AU       Queensland   admin   

   population  population_proper  newCol  newCol1  
0     5312163            4840600  newVal     1035  
1     5078193            4529500  newVal     1034  
2     2514184            2360241  newVal     1033

In [109]:
# one way of selecting the name of the first city reviewed from each admin_name in the dataset
readd.groupby('admin_name').apply(lambda df: df.city.iloc[0])

admin_name
Australian Capital Territory     Canberra
New South Wales                    Sydney
Northern Territory                 Darwin
Queensland                       Brisbane
South Australia                  Adelaide
Tasmania                           Hobart
Victoria                        Melbourne
Western Australia                   Perth
dtype: object

In [110]:
# How to pick out the top population by admin_name and city: (Multi-index)
readd.groupby(['admin_name', 'city']).apply(lambda df: df.loc[df.population_proper.idxmax()])

city      lat       lng  \
admin_name                   city                                      
Australian Capital Territory Canberra    Canberra -35.2931  149.1269   
New South Wales              Aberdare    Aberdare -32.8428  151.3803   
                             Aberdeen    Aberdeen -32.1650  150.9011   
                             Abermain    Abermain -32.8072  151.4275   
                             Albury        Albury -36.0806  146.9158   
...                                           ...      ...       ...   
Western Australia            Wundowie    Wundowie -31.7667  116.3830   
                             Yallingup  Yallingup -33.6397  115.0263   
                             Yanchep      Yanchep -31.5500  115.6340   
                             York            York -31.8883  116.7686   
                             Yunderup    Yunderup -32.5833  115.8000   

                                          country iso2  \
admin_name                   city                        
Australian Capital Territory Canberra   Australia   AU   
New South Wales              Aberdare   Australia   AU   
                             Aberdeen   Australia   AU   
                             Abermain   Australia   AU   
                             Albury     Australia   AU   
...                                           ...  ...   
Western Australia            Wundowie   Australia   AU   
                             Yallingup  Australia   AU   
                             Yanchep    Australia   AU   
                             York       Australia   AU   
                             Yunderup   Australia   AU   

                                                          admin_name  capital  \
admin_name                   city                                               
Australian Capital Territory Canberra   Australian Capital Territory  primary   
New South Wales              Aberdare                New South Wales      NaN   
                             Aberdeen                New South Wales      NaN   
                             Abermain                New South Wales      NaN   
                             Albury                  New South Wales      NaN   
...                                                              ...      ...   
Western Australia            Wundowie              Western Australia      NaN   
                             Yallingup             Western Australia      NaN   
                             Yanchep               Western Australia      NaN   
                             York                  Western Australia      NaN   
                             Yunderup              Western Australia      NaN   

                                        population  population_proper  newCol  \
admin_name                   city                                               
Australian Capital Territory Canberra       426704             381488  newVal   
New South Wales              Aberdare         2473               2473  newVal   
                             Aberdeen         2084               2084  newVal   
                             Abermain         2337               2337  newVal   
                             Albury          47974               4804  newVal   
...                                            ...                ...     ...   
Western Australia            Wundowie         1352               1352  newVal   
                             Yallingup        1029               1029  newVal   
                             Yanchep          8868               8868  newVal   
                             York             2548               2535  newVal   
                             Yunderup         3114               3114  newVal   

                                        newCol1  
admin_name                   city                
Australian Capital Territory Canberra      1029  
New South Wales              Aberdare       494  
                             Aberdeen       4

In [113]:
# Generate a simple statistical summary of the dataset
readd.groupby(['admin_name']).population_proper.agg([len, min, max])

len     min      max
admin_name                                        
Australian Capital Territory    1  381488   381488
New South Wales               335     180  4840600
Northern Territory             10    1081   136828
Queensland                    240     118  2360241
South Australia                88     195  1295714
Tasmania                       46     768   222356
Victoria                      232     865  4529500
Western Australia              83    1029  2039200

Sorting
--

In [114]:
# We can use sort_values() to sort based on values
readd.sort_values(by='admin_name')

city      lat       lng    country iso2  \
6          Canberra -35.2931  149.1269  Australia   AU   
0            Sydney -33.8650  151.2094  Australia   AU   
1033        Tomakin -35.8228  150.1894  Australia   AU   
514          Lawson -33.7178  150.4322  Australia   AU   
512         Maclean -29.4567  153.2011  Australia   AU   
...             ...      ...       ...        ...  ...   
687   Kambalda West -31.2114  121.7129  Australia   AU   
325        Narrogin -32.9360  117.1780  Australia   AU   
35        Geraldton -28.7744  114.6089  Australia   AU   
589      Ravenswood -32.5833  115.8330  Australia   AU   
506        Wooroloo -31.8030  116.3140  Australia   AU   

                        admin_name  capital  population  population_proper  \
6     Australian Capital Territory  primary      426704             381488   
0                  New South Wales    admin     5312163            4840600   
1033               New South Wales      NaN        1001               1001   
514                New South Wales      NaN        2614               2614   
512                New South Wales      NaN        2628               2628   
...                            ...      ...         ...                ...   
687              Western Australia      NaN        1789               1789   
325              Western Australia      NaN        4274               4274   
35               Western Australia      NaN       37648              37648   
589              Western Australia      NaN        2176               2176   
506              Western Australia      NaN        2642               2642   

      newCol  newCol1  
6     newVal     1029  
0     newVal     1035  
1033  newVal        2  
514   newVal      521  
512   newVal      523  
...      ...      ...  
687   newVal      348  
325   newVal      710  
35    newVal     1000  
589   newVal      446  
506   newVal      529  

[1035 rows x 11 columns]

In [116]:
# By default, sort happens in ascending order, to get in descending order, we need to do as follows
readd.sort_values(by='admin_name', ascending=False)

city      lat       lng    country iso2  \
92    Australind -33.2792  115.7150  Australia   AU   
461  Leschenault -33.2450  115.7350  Australia   AU   
948  Bakers Hill -31.7510  116.4550  Australia   AU   
479   Donnybrook -33.5800  115.8200  Australia   AU   
473     Merredin -31.4820  118.2790  Australia   AU   
..           ...      ...       ...        ...  ...   
489       Uralla -30.6472  151.4956  Australia   AU   
488    Coonamble -30.9317  148.3678  Australia   AU   
485      Howlong -35.9817  146.6344  Australia   AU   
517      Narooma -36.2000  150.1333  Australia   AU   
6       Canberra -35.2931  149.1269  Australia   AU   

                       admin_name  capital  population  population_proper  \
92              Western Australia      NaN       14539              14539   
461             Western Australia      NaN        2932               2932   
948             Western Australia      NaN        1118               1118   
479             Western Australia      NaN        2824               2824   
473             Western Australia      NaN        2850               2850   
..                            ...      ...         ...                ...   
489               New South Wales      NaN        2743               2743   
488               New South Wales      NaN        2750               2750   
485               New South Wales      NaN        2777               2777   
517               New South Wales      NaN        2605               2605   
6    Australian Capital Territory  primary      426704             381488   

     newCol  newCol1  
92   newVal      943  
461  newVal      574  
948  newVal       87  
479  newVal      556  
473  newVal      562  
..      ...      ...  
489  newVal      546  
488  newVal      547  
485  newVal      550  
517  newVal      518  
6    newVal     1029  

[1035 rows x 11 columns]

In [118]:
# To sort by index, we use sort_index()
readd.sort_index()

city      lat       lng    country iso2         admin_name capital  \
0        Sydney -33.8650  151.2094  Australia   AU    New South Wales   admin   
1     Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
2      Brisbane -27.4678  153.0281  Australia   AU         Queensland   admin   
3         Perth -31.9522  115.8589  Australia   AU  Western Australia   admin   
4      Adelaide -34.9289  138.6011  Australia   AU    South Australia   admin   
...         ...      ...       ...        ...  ...                ...     ...   
1030  The Palms -26.1902  152.5861  Australia   AU         Queensland     NaN   
1031  Millfield -32.8886  151.2614  Australia   AU    New South Wales     NaN   
1032   Dunbogan -31.6500  152.8167  Australia   AU    New South Wales     NaN   
1033    Tomakin -35.8228  150.1894  Australia   AU    New South Wales     NaN   
1034     Currie -39.9311  143.8510  Australia   AU           Tasmania   minor   

      population  population_proper  newCol  newCol1  
0        5312163            4840600  newVal     1035  
1        5078193            4529500  newVal     1034  
2        2514184            2360241  newVal     1033  
3        2059484            2039200  newVal     1032  
4        1345777            1295714  newVal     1031  
...          ...                ...     ...      ...  
1030        1010               1010  newVal        5  
1031        1006               1006  newVal        4  
1032        1002               1002  newVal        3  
1033        1001               1001  newVal        2  
1034         768                768  newVal        1  

[1035 rows x 11 columns]

In [119]:
# We can sort based on one or more columns
readd.sort_values(by=['admin_name', 'population_proper'])

city      lat       lng    country iso2  \
6       Canberra -35.2931  149.1269  Australia   AU   
31      Tamworth -31.0833  150.9167  Australia   AU   
88       Kempsey -31.0833  152.8333  Australia   AU   
1033     Tomakin -35.8228  150.1894  Australia   AU   
1032    Dunbogan -31.6500  152.8167  Australia   AU   
...          ...      ...       ...        ...  ...   
43        Albany -35.0228  117.8814  Australia   AU   
48    Kalgoorlie -30.7489  121.4658  Australia   AU   
35     Geraldton -28.7744  114.6089  Australia   AU   
22       Bunbury -33.3333  115.6333  Australia   AU   
3          Perth -31.9522  115.8589  Australia   AU   

                        admin_name  capital  population  population_proper  \
6     Australian Capital Territory  primary      426704             381488   
31                 New South Wales      NaN       42872                180   
88                 New South Wales      NaN       15309                474   
1033               New South Wales      NaN        1001               1001   
1032               New South Wales      NaN        1002               1002   
...                            ...      ...         ...                ...   
43               Western Australia      NaN       34205              29373   
48               Western Australia      NaN       29849              29849   
35               Western Australia      NaN       37648              37648   
22               Western Australia      NaN       74363              71090   
3                Western Australia    admin     2059484            2039200   

      newCol  newCol1  
6     newVal     1029  
31    newVal     1004  
88    newVal      947  
1033  newVal        2  
1032  newVal        3  
...      ...      ...  
43    newVal      992  
48    newVal      987  
35    newVal     1000  
22    newVal     1013  
3     newVal     1032  

[1035 rows x 11 columns]

Finding datatype of a column
--

In [121]:
# To find data type of a column
readd.admin_name.dtype

dtype('O')

In [124]:
# To get the data type of whole dataset columns
readd.dtypes

city                  object
lat                  float64
lng                  float64
country               object
iso2                  object
admin_name            object
capital               object
population             int64
population_proper      int64
newCol                object
newCol1                int64
dtype: object

In [129]:
# To convert a data type to another
readd.newCol1.astype('float64')

0       1035.0
1       1034.0
2       1033.0
3       1032.0
4       1031.0
         ...  
1030       5.0
1031       4.0
1032       3.0
1033       2.0
1034       1.0
Name: newCol1, Length: 1035, dtype: float64

Missing Data
--

In [141]:
# To select NaN entries if any - For technical reasons these NaN values are always of the 'float64' dtype.
readd[pd.isnull(readd.capital)]

# Other ways to find missing values
# missing_price_reviews = reviews[reviews.price.isnull()]
# n_missing_prices = len(missing_price_reviews)
# Cute alternative solution: if we sum a boolean series, True is treated as 1 and False as 0
# n_missing_prices = reviews.price.isnull().sum()
# or equivalently:
# n_missing_prices = pd.isnull(reviews.price).sum()


city      lat       lng    country iso2       admin_name capital  \
5     Gold Coast -28.0167  153.4000  Australia   AU       Queensland     NaN   
7      Newcastle -32.9167  151.7500  Australia   AU  New South Wales     NaN   
8     Wollongong -34.4331  150.8831  Australia   AU  New South Wales     NaN   
9        Geelong -38.1500  144.3500  Australia   AU         Victoria     NaN   
11    Townsville -19.2564  146.8183  Australia   AU       Queensland     NaN   
...          ...      ...       ...        ...  ...              ...     ...   
1029     Capella -23.0667  148.0167  Australia   AU       Queensland     NaN   
1030   The Palms -26.1902  152.5861  Australia   AU       Queensland     NaN   
1031   Millfield -32.8886  151.2614  Australia   AU  New South Wales     NaN   
1032    Dunbogan -31.6500  152.8167  Australia   AU  New South Wales     NaN   
1033     Tomakin -35.8228  150.1894  Australia   AU  New South Wales     NaN   

      population  population_proper  newCol  newCol1  
5         679127             638090  newVal     1030  
7         322278             152984  newVal     1028  
8         302739             289236  newVal     1027  
9         263280             191440  newVal     1026  
11        180820             178649  newVal     1024  
...          ...                ...     ...      ...  
1029        1010               1010  newVal        6  
1030        1010               1010  newVal        5  
1031        1006               1006  newVal        4  
1032        1002               1002  newVal        3  
1033        1001               1001  newVal        2  

[1026 rows x 11 columns]

In [149]:
# To select not NaN entries if any
readd[pd.notnull(readd.capital)]

city      lat       lng    country iso2  \
0        Sydney -33.8650  151.2094  Australia   AU   
1     Melbourne -37.8136  144.9631  Australia   AU   
2      Brisbane -27.4678  153.0281  Australia   AU   
3         Perth -31.9522  115.8589  Australia   AU   
4      Adelaide -34.9289  138.6011  Australia   AU   
6      Canberra -35.2931  149.1269  Australia   AU   
10       Hobart -42.8806  147.3250  Australia   AU   
15       Darwin -12.4381  130.8411  Australia   AU   
1034     Currie -39.9311  143.8510  Australia   AU   

                        admin_name  capital  population  population_proper  \
0                  New South Wales    admin     5312163            4840600   
1                         Victoria    admin     5078193            4529500   
2                       Queensland    admin     2514184            2360241   
3                Western Australia    admin     2059484            2039200   
4                  South Australia    admin     1345777            1295714   
6     Australian Capital Territory  primary      426704             381488   
10                        Tasmania    admin      240342             222356   
15              Northern Territory    admin      136828             136828   
1034                      Tasmania    minor         768                768   

      newCol  newCol1  
0     newVal     1035  
1     newVal     1034  
2     newVal     1033  
3     newVal     1032  
4     newVal     1031  
6     newVal     1029  
10    newVal     1025  
15    newVal     1020  
1034  newVal        1

In [143]:
# If any NaN values present, we could replace those values using 'fillna()'
readd.capital.fillna("Unknown")

0         admin
1         admin
2         admin
3         admin
4         admin
         ...   
1030    Unknown
1031    Unknown
1032    Unknown
1033    Unknown
1034      minor
Name: capital, Length: 1035, dtype: object

In [150]:
# We can replace a column value to other using 'replace()'
readd.admin_name.replace('New South Wales', 'NSW')

0                     NSW
1                Victoria
2              Queensland
3       Western Australia
4         South Australia
              ...        
1030           Queensland
1031                  NSW
1032                  NSW
1033                  NSW
1034             Tasmania
Name: admin_name, Length: 1035, dtype: object

Renaming and Combining
--

In [151]:
# Renaming a column
readd.rename(columns={'admin_name': 'State'})

city      lat       lng    country iso2              State capital  \
0        Sydney -33.8650  151.2094  Australia   AU    New South Wales   admin   
1     Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
2      Brisbane -27.4678  153.0281  Australia   AU         Queensland   admin   
3         Perth -31.9522  115.8589  Australia   AU  Western Australia   admin   
4      Adelaide -34.9289  138.6011  Australia   AU    South Australia   admin   
...         ...      ...       ...        ...  ...                ...     ...   
1030  The Palms -26.1902  152.5861  Australia   AU         Queensland     NaN   
1031  Millfield -32.8886  151.2614  Australia   AU    New South Wales     NaN   
1032   Dunbogan -31.6500  152.8167  Australia   AU    New South Wales     NaN   
1033    Tomakin -35.8228  150.1894  Australia   AU    New South Wales     NaN   
1034     Currie -39.9311  143.8510  Australia   AU           Tasmania   minor   

      population  population_proper  newCol  newCol1  
0        5312163            4840600  newVal     1035  
1        5078193            4529500  newVal     1034  
2        2514184            2360241  newVal     1033  
3        2059484            2039200  newVal     1032  
4        1345777            1295714  newVal     1031  
...          ...                ...     ...      ...  
1030        1010               1010  newVal        5  
1031        1006               1006  newVal        4  
1032        1002               1002  newVal        3  
1033        1001               1001  newVal        2  
1034         768                768  newVal        1  

[1035 rows x 11 columns]

In [152]:
# Renaming the index
readd.rename(index={0: 'zero', 1: 'one'})

# You'll probably rename columns very often, but rename index values very rarely. 
# For that, 'set_index()' is usually more convenient.

city      lat       lng    country iso2         admin_name capital  \
zero     Sydney -33.8650  151.2094  Australia   AU    New South Wales   admin   
one   Melbourne -37.8136  144.9631  Australia   AU           Victoria   admin   
2      Brisbane -27.4678  153.0281  Australia   AU         Queensland   admin   
3         Perth -31.9522  115.8589  Australia   AU  Western Australia   admin   
4      Adelaide -34.9289  138.6011  Australia   AU    South Australia   admin   
...         ...      ...       ...        ...  ...                ...     ...   
1030  The Palms -26.1902  152.5861  Australia   AU         Queensland     NaN   
1031  Millfield -32.8886  151.2614  Australia   AU    New South Wales     NaN   
1032   Dunbogan -31.6500  152.8167  Australia   AU    New South Wales     NaN   
1033    Tomakin -35.8228  150.1894  Australia   AU    New South Wales     NaN   
1034     Currie -39.9311  143.8510  Australia   AU           Tasmania   minor   

      population  population_proper  newCol  newCol1  
zero     5312163            4840600  newVal     1035  
one      5078193            4529500  newVal     1034  
2        2514184            2360241  newVal     1033  
3        2059484            2039200  newVal     1032  
4        1345777            1295714  newVal     1031  
...          ...                ...     ...      ...  
1030        1010               1010  newVal        5  
1031        1006               1006  newVal        4  
1032        1002               1002  newVal        3  
1033        1001               1001  newVal        2  
1034         768                768  newVal        1  

[1035 rows x 11 columns]

In [153]:
# Renaming both the row and column index
readd.rename_axis("Rows", axis='rows').rename_axis("Columns", axis='columns')

Columns       city      lat       lng    country iso2         admin_name  \
Rows                                                                       
0           Sydney -33.8650  151.2094  Australia   AU    New South Wales   
1        Melbourne -37.8136  144.9631  Australia   AU           Victoria   
2         Brisbane -27.4678  153.0281  Australia   AU         Queensland   
3            Perth -31.9522  115.8589  Australia   AU  Western Australia   
4         Adelaide -34.9289  138.6011  Australia   AU    South Australia   
...            ...      ...       ...        ...  ...                ...   
1030     The Palms -26.1902  152.5861  Australia   AU         Queensland   
1031     Millfield -32.8886  151.2614  Australia   AU    New South Wales   
1032      Dunbogan -31.6500  152.8167  Australia   AU    New South Wales   
1033       Tomakin -35.8228  150.1894  Australia   AU    New South Wales   
1034        Currie -39.9311  143.8510  Australia   AU           Tasmania   

Columns capital  population  population_proper  newCol  newCol1  
Rows                                                             
0         admin     5312163            4840600  newVal     1035  
1         admin     5078193            4529500  newVal     1034  
2         admin     2514184            2360241  newVal     1033  
3         admin     2059484            2039200  newVal     1032  
4         admin     1345777            1295714  newVal     1031  
...         ...         ...                ...     ...      ...  
1030        NaN        1010               1010  newVal        5  
1031        NaN        1006               1006  newVal        4  
1032        NaN        1002               1002  newVal        3  
1033        NaN        1001               1001  newVal        2  
1034      minor         768                768  newVal        1  

[1035 rows x 11 columns]

Combining
--

In [ ]:
# Read two datasets and combine them
canadian_youtube = pd.read_csv("filename.csv")
british_youtube = pd.read_csv("filename.csv")

pd.concat([canadian_youtube, british_youtube])

In [ ]:
# 'join()' lets you combine different DataFrame objects which have an index in common

#To pull down videos that happened to be trending on the same day in both Canada and the UK, we could do the following:

left = canadian_youtube.set_index(['title', 'trending_date'])
right = british_youtube.set_index(['title', 'trending_date'])

left.join(right, lsuffix='_CAN', rsuffix='_UK')

# The lsuffix and rsuffix parameters are necessary here because the data has the same column names in both British and Canadian datasets. 
# If this wasn't true (because, say, we'd renamed them beforehand) we wouldn't need them.